File: 04-interfaces/main.go
package main

import "fmt"

// NestJS 里类似：定义接口类型，不同 Service 实现同一契约

type Storer interface {
	Save(title string) error
}

type MemoryStore struct {
	notes []string
}

func (m *MemoryStore) Save(title string) error {
	m.notes = append(m.notes, title)
	return nil
}

func (m *MemoryStore) List() []string {
	return m.notes
}

// FileStore 故意不实现 List，只实现 Storer 也能赋给 interface

type FileStore struct{}

func (f *FileStore) Save(title string) error {
	fmt.Println("[file] save:", title)
	return nil
}

func createNote(s Storer, title string) error {
	return s.Save(title)
}

func main() {
	mem := &MemoryStore{}
	_ = createNote(mem, "第一条")
	_ = createNote(mem, "第二条")
	fmt.Println("memory:", mem.List())

	file := &FileStore{}
	_ = createNote(file, "写入文件（示例打印）")

	// 类型断言：只有确定具体类型时才用
	if m, ok := interface{}(mem).(*MemoryStore); ok {
		fmt.Println("断言成功，条数:", len(m.List()))
	}
}

File: 04-interfaces/nodejs/main.ts
// Go: interface 隐式实现 + 类型断言；TS 用 interface + class + instanceof

interface Storer {
  save(title: string): void;
}

class MemoryStore implements Storer {
  private notes: string[] = [];

  save(title: string): void {
    this.notes.push(title);
  }

  list(): string[] {
    return this.notes;
  }
}

// 只实现 Storer，不实现 list — 与 Go FileStore 一致
class FileStore implements Storer {
  save(title: string): void {
    console.log('[file] save:', title);
  }
}

function createNote(s: Storer, title: string): void {
  s.save(title);
}

function main(): void {
  const mem = new MemoryStore();
  createNote(mem, '第一条');
  createNote(mem, '第二条');
  console.log('memory:', mem.list());

  const file = new FileStore();
  createNote(file, '写入文件（示例打印）');

  // 类型收窄：对应 Go 的 m, ok := x.(*MemoryStore)
  if (mem instanceof MemoryStore) {
    console.log('断言成功，条数:', mem.list().length);
  }
}

main();
